GafferScene context sanitisation

johnhaddon opened the pull request:
This improves scene generation performance by reducing the number of unique contexts created, which in turn reduces pressure on the hash cache, allowing us to use less memory and retain a more useful working set. For a particularly troubling scene I'm working with, this has almost halved scene generation time and memory usage.

Recap : when we're evaluating a scene, we use the following special context variables :

- "scene:path" : this specifies which location within the scene we want to evaluate, when computing bounds, transforms, attributes, objects and child names. This can easily range in the tens to hundreds of thousands of unique values.
- "scene:setName" : this specifies the name of the set we want to evaluate when pulling on the set plug. It can easily run into the hundreds of unique values.
- "scene:filter:inputScene" : this specifies which input scene a filter should evaluate for. This can easily run into the hundreds, if not more.

The key here is that the number of unique values of each of the variables in isolation is actually pretty reasonable, and shouldn't pose a huge problem for the cache. But if we allow permutations to exist, we can easily run into trouble. And permutations aren't actually useful - for any given evaluation, only one of the variables is needed.

Here we introduce a ContextSanitiser class which can be enabled via the `stats -contextSanitiser` argument, and which outputs warnings whenever a variable has been specified unnecessarily, potentially leading to degraded cache performance. It then fixes every such "leak" in all GafferScene nodes, and ensures we can't get future problems by activating the ContextSanitiser whenever we run the unit tests.

As with the hash cache revamp, I'm just formalising work done by others here - @danieldresser-ie and @davidsminor were instrumental in figuring all this out in the first place, and I'm late to the party in realising that it needs to be treated as a central priority.

danieldresser-ie commented:
The context sanitiser seems pretty good - it helped while tracking down the globals leak with the IE denoiser.

With all the little changes to improve contexts, I didn't study them all closely enough to confirm that there isn't some other piece of code that depends on the context that you've fixed, but the changes all make sense.

I guess if the tests pass, then LGTM.

( Andrew did point out that we still haven't addressed the SceneReader fileName depending on scene:path, but that's not a standard plug type, so we can't catch it here I guess ).

johnhaddon commented:
> Andrew did point out that we still haven't addressed the SceneReader fileName depending on scene:path, but that's not a standard plug type, so we can't catch it here I guess.

I felt that "make sanitiser and sanitise all the problems it finds" was a sensible unit of work for a single PR. The SceneReader thing is one of several avenues I will be exploring next...

andrewkaufman commented:
I guess it’s a little late, but there were some TimeWarp changes bundled into other commits. Was that intentional or accidental?

Also, did you give any consideration to sanitizing non-scene variables? I’ve seen a couple scripts where the slap comp is quite slow because it’s pulling on the scene globals via an expression, with image tile/channel context variables causing rehashes.

johnhaddon commented:
> I guess it’s a little late, but there were some TimeWarp changes bundled into other commits. Was that intentional or accidental?

Intentional. In da6f37dc26ba27debef3d282c3f8ad364e0afe6d I removed some awkward juggling that was made unnecessary by the main change to Context::EditableScope. In 350939984f06312fa8f4337860ec6932fce42fda I removed a GlobalScope that was now being provided by the ContextProcessor base class.

> Also, did you give any consideration to sanitizing non-scene variables? I’ve seen a couple scripts where the slap comp is quite slow because it’s pulling on the scene globals via an expression, with image tile/channel context variables causing rehashes.

I definitely planned to make a `GafferImageTest::ContextSanitiser` class in future, and do the equivalent sanitisation work across GafferImage. I asked @danieldresser-ie for a good test case but I don't have one yet. I hadn't considered the image->scene cross-contamination you pointed out, but that sounds like a good avenue for exploration too. When we do #1665 I'm anticipating that the scene query nodes will be responsible for sanitising the query contexts, because we don't have control over where a user might use an expression, and that could take care of some of that.

My current focus though is on taming a couple of nasty production setups that require hash cache limits that are higher than we're comfortable with. I have a few avenues for exploration there that are unrelated to the above, so I'm giving those priority for now...
